File: js/src/completion.ts
import { CodeCompletionCore } from "antlr4-c3";
import { CommonTokenStream } from "antlr4ng";

import { CaseChangingCharStream } from "./caseChangingCharStream.js";
import { CallsLexer } from "./generated/CallsLexer.js";
import { CallsParser } from "./generated/CallsParser.js";
import tokens from "./generated/canonical-tokens.json" with { type: "json" };

const WORDS: Record<string, string> = tokens.words;
const CATEGORIES: Record<string, string> = tokens.categories;

/**
 * A run of words that is one call between them -- "Full Auto", "Shrug Off",
 * "Receding Tide". The grammar sees each of these as several keywords, but a
 * player shouts one call, so autocomplete offers and accepts them whole.
 */
interface Phrase {
  /** Token names, in order. */
  names: string[];
  /** Canonical words, in order -- `WORDS[names[i]]`. */
  words: string[];
  /** The whole phrase as one label, e.g. "Full Auto". */
  label: string;
  /** Coarse grouping, one of `categoryOrder`. */
  category: string;
  /**
   * Whether this is a Defense name. Names are admitted by a rule rather than by
   * their keywords (`defenseWord : ~NUMBER`), so a different question decides
   * whether one may go at the caret.
   */
  isName: boolean;
}

function phrase(names: string[], isName: boolean): Phrase {
  const words = names.map((name) => WORDS[name]);
  return {
    names,
    words,
    label: words.join(" "),
    // A name's words keep their own categories elsewhere -- FULL is an
    // "effect", because of "Full Auto" -- but a Defense name is a defense name
    // and nothing else, which is what keeps "Full" from leaking out of the
    // Mitigate hint as an effect. Every other phrase takes its first word's.
    category: isName ? "defense-name" : CATEGORIES[names[0]],
    isName,
  };
}

/** The keyword runs the grammar joins: `FULL AUTO`, `SHRUG OFF`, `PHASE (OUT | IN)`. */
const PHRASES: Phrase[] = tokens.phrases.map((names) => phrase(names, false));

/**
 * The Defense names the rulebook does list, as whole names.
 *
 * The grammar accepts any word at all as a Defense name (`defenseWord :
 * ~NUMBER`) and doesn't require "Receding" to be followed by "Tide", so there is
 * nothing in it left to complete against -- these are offered on its behalf.
 * Kept in shared/canonical-tokens.json rather than here so the Python engine
 * offers the same set.
 */
const DEFENSE_NAMES: Phrase[] = tokens.defenseNames.map((names) => phrase(names, true));

const ALL_PHRASES: Phrase[] = [...PHRASES, ...DEFENSE_NAMES];

/**
 * How many words of a phrase the caret may already have behind it and still
 * complete the whole thing, so that "Shrug O" offers "Shrug Off" rather than
 * just "Off". Derived from the tables rather than fixed at one, so a
 * three-word call would need no change here.
 */
const MAX_WORDS_TYPED = Math.max(...ALL_PHRASES.map((p) => p.names.length)) - 1;

/**
 * What the player has typed, in the form a phrase label can be compared
 * against: lower-cased, with every run of separators collapsed to the single
 * space labels are written with, so "Full-A" and "shrug,of" match too.
 */
function typedPrefix(slice: string): string {
  return slice.toLowerCase().replace(/[ \t\r\n\-,:]+/g, " ");
}

export interface KeywordSuggestion {
  kind: "keyword";
  /** The whole call to insert, e.g. "Shrug Off" -- a single word for most calls. */
  label: string;
  /**
   * Just the word the caret is completing: "Off" for a caret after "Shrug ",
   * "Shrug" for a caret at the start of input. What a hint means when it names
   * the words allowed at a position.
   */
  word: string;
  /** Coarse grouping, one of `categoryOrder`. */
  category: string;
  /**
   * Half-open range in the input that `label` replaces. It reaches back over
   * any words of the phrase already typed, so accepting a suggestion never
   * leaves a partial word behind.
   */
  start: number;
  end: number;
}

export interface NumberSuggestion {
  kind: "number";
  /** Half-open range in the input a number would occupy. */
  start: number;
  end: number;
}

export type Suggestion = KeywordSuggestion | NumberSuggestion;

function findCaretTokenIndex(tokenStream: CommonTokenStream, cursor: number): number {
  const all = tokenStream.getTokens();
  for (let i = 0; i < all.length; i++) {
    if (cursor <= all[i].stop + 1) {
      return i;
    }
  }
  return all.length - 1;
}

/**
 * Suggest completions for `text` at caret offset `cursor` (a UTF-16 code
 * unit index). Uses antlr4-c3's ATN-based CodeCompletionCore, which is the
 * reason the JS/TS side is built on antlr4ng rather than the classic
 * antlr4 runtime -- antlr4-c3 only supports antlr4ng.
 *
 * Multi-word calls come back as one suggestion each. The candidate set antlr4-c3
 * reports is per token, so "Shrug Off" would otherwise arrive as "Shrug" and
 * then "Off"; the phrase tables above say which tokens belong to one call, and
 * every suggestion carries the input range it replaces so that accepting one
 * mid-phrase rewrites the words already typed instead of appending to them.
 */
export function suggest(text: string, cursor: number): Suggestion[] {
  const stream = new CaseChangingCharStream(text, true);
  const lexer = new CallsLexer(stream);
  const tokenStream = new CommonTokenStream(lexer);
  tokenStream.fill();

  const parser = new CallsParser(tokenStream);
  parser.removeErrorListeners();
  // Tolerant parse: default error-recovery builds a best-effort tree even
  // over invalid/partial input, which is what collectCandidates needs.
  parser.call();

  const caretTokenIndex = findCaretTokenIndex(tokenStream, cursor);

  const core = new CodeCompletionCore(parser);
  core.ignoredTokens = new Set([CallsLexer.EOF]);
  // `defenseWord` matches any token but NUMBER, so left alone the core would
  // report every keyword in the language wherever a Defense name may go.
  // Marking it preferred stops the walk at the rule instead, which is answered
  // with DEFENSE_NAMES below.
  core.preferredRules = new Set([CallsParser.RULE_defenseWord]);
  const candidates = core.collectCandidates(caretTokenIndex);

  let numberAllowed = false;
  const allowedNames = new Set<string>();
  for (const tokenType of candidates.tokens.keys()) {
    if (tokenType === CallsLexer.NUMBER) {
      numberAllowed = true;
      continue;
    }
    const name = CallsLexer.symbolicNames[tokenType];
    if (name !== null && name !== undefined) {
      allowedNames.add(name);
    }
  }
  const nameAllowed = candidates.rules.has(CallsParser.RULE_defenseWord);

  const all = tokenStream.getTokens();
  const suggestions: Suggestion[] = [];
  const seenLabels = new Set<string>();
  /** Words already spoken for by a phrase, so the bare word isn't offered too. */
  const phraseWords = new Set<string>();

  // Phrases first, and from the furthest-back anchor forwards: a phrase that
  // reaches back over words the player has already typed beats the bare next
  // word, and the earliest anchor that still matches is the one whose range
  // rewrites the whole phrase rather than just its last word.
  for (
    let anchor = Math.max(0, caretTokenIndex - MAX_WORDS_TYPED);
    anchor <= caretTokenIndex;
    anchor++
  ) {
    const wordsTyped = caretTokenIndex - anchor;
    const start = all[anchor].start;
    const typed = typedPrefix(text.slice(start, cursor));
    for (const candidate of ALL_PHRASES) {
      if (wordsTyped >= candidate.names.length) continue;
      // Whether the word the caret is on may go here at all. Indexing the
      // phrase by how many of its words are already typed is what makes one
      // test serve both "Ful" (nothing typed yet, so FULL must be allowed) and
      // "Full " (one word in, so AUTO must be).
      const allowed = candidate.isName
        ? nameAllowed
        : allowedNames.has(candidate.names[wordsTyped]);
      if (!allowed) continue;
      // Everything from the anchor to the caret has to read as the start of
      // this phrase -- which also confirms the earlier words really are its
      // earlier words, so no separate token-by-token check is needed.
      if (!candidate.label.toLowerCase().startsWith(typed)) continue;
      if (seenLabels.has(candidate.label)) continue;
      seenLabels.add(candidate.label);
      phraseWords.add(candidate.words[wordsTyped]);
      suggestions.push({
        kind: "keyword",
        label: candidate.label,
        word: candidate.words[wordsTyped],
        category: candidate.category,
        start,
        end: cursor,
      });
    }
  }

  // Then the plain keywords, which only ever complete the word the caret is
  // actually on.
  const caretStart = all[caretTokenIndex].start;
  const caretTyped = typedPrefix(text.slice(caretStart, cursor));
  for (const name of allowedNames) {
    const label = WORDS[name];
    // Canonical words are capitalized; what the user has typed so far is
    // whatever they typed, lower-cased above -- so match case-insensitively
    // while still offering the canonical spelling.
    if (label === undefined || !label.toLowerCase().startsWith(caretTyped)) continue;
    // "Auto" on its own is never the answer when "Full Auto" is already on
    // offer, and neither is "Full" when it can only begin one.
    if (phraseWords.has(label) || seenLabels.has(label)) continue;
    seenLabels.add(label);
    suggestions.push({
      kind: "keyword",
      label,
      word: label,
      category: CATEGORIES[name],
      start: caretStart,
      end: cursor,
    });
  }

  if (numberAllowed && /^[0-9]*$/.test(caretTyped)) {
    suggestions.push({ kind: "number", start: caretStart, end: cursor });
  }

  suggestions.sort((a, b) => {
    const aKey = a.kind === "number" ? "" : a.label;
    const bKey = b.kind === "number" ? "" : b.label;
    return aKey.localeCompare(bKey);
  });
  return suggestions;
}

File: python/src/altered_calls_parser/_completion.py
"""Autocomplete via Brzozowski derivatives over the declarative grammar
mirror in _grammar_shape.py. This mirrors what antlr4-c3 does by walking
an ATN, but there is no Python port of antlr4-c3, and the grammar is
small/flat/non-recursive enough for a hand-rolled derivative walk to be
tractable and easy to keep correct.
"""

from dataclasses import dataclass

from antlr4 import CommonTokenStream

from ._case_changing_stream import CaseChangingCharStream
from ._grammar_shape import (
    ANY_WORD_TOKEN,
    CALL,
    NUMBER_TOKEN,
    Alt,
    AnyWordSlot,
    Lit,
    Node,
    NumberSlot,
    Opt,
    Rep,
    Seq,
)
from ._tokens import DEFENSE_NAME_WORDS, WORDS
from .generated.CallsLexer import CallsLexer


class _Empty:
    """Matches only the empty remainder; nullable, consumes nothing further."""


class _Fail:
    """A dead end; matches nothing."""


EMPTY = _Empty()
FAIL = _Fail()


def nullable(node: Node | _Empty | _Fail) -> bool:
    if isinstance(node, _Empty):
        return True
    if isinstance(node, _Fail):
        return False
    if isinstance(node, (Lit, NumberSlot, AnyWordSlot)):
        return False
    if isinstance(node, Opt):
        return True
    if isinstance(node, Rep):
        # One or more, so an empty match is never enough.
        return False
    if isinstance(node, Seq):
        return all(nullable(p) for p in node.parts)
    if isinstance(node, Alt):
        return any(nullable(o) for o in node.options)
    raise TypeError(node)


def first(node: Node | _Empty | _Fail) -> set[str]:
    if isinstance(node, (_Empty, _Fail)):
        return set()
    if isinstance(node, Lit):
        return {node.name}
    if isinstance(node, NumberSlot):
        return {NUMBER_TOKEN}
    if isinstance(node, AnyWordSlot):
        return {ANY_WORD_TOKEN}
    if isinstance(node, (Opt, Rep)):
        return first(node.part)
    if isinstance(node, Seq):
        result: set[str] = set()
        for part in node.parts:
            result |= first(part)
            if not nullable(part):
                break
        return result
    if isinstance(node, Alt):
        result = set()
        for option in node.options:
            result |= first(option)
        return result
    raise TypeError(node)


def _seq_of(parts: tuple[Node | _Empty | _Fail, ...]) -> Node | _Empty | _Fail:
    filtered = [p for p in parts if not isinstance(p, _Empty)]
    if not filtered:
        return EMPTY
    if any(isinstance(p, _Fail) for p in filtered):
        return FAIL
    if len(filtered) == 1:
        return filtered[0]
    return Seq(tuple(filtered))  # type: ignore[arg-type]


def _alt_of(branches: list[Node | _Empty | _Fail]) -> Node | _Empty | _Fail:
    non_fail = [b for b in branches if not isinstance(b, _Fail)]
    if not non_fail:
        return FAIL
    if len(non_fail) == 1:
        return non_fail[0]
    return Alt(tuple(non_fail))  # type: ignore[arg-type]


def derivative(node: Node | _Empty | _Fail, token: str) -> Node | _Empty | _Fail:
    if isinstance(node, (_Empty, _Fail)):
        return FAIL
    if isinstance(node, Lit):
        return EMPTY if node.name == token else FAIL
    if isinstance(node, NumberSlot):
        return EMPTY if token == NUMBER_TOKEN else FAIL
    if isinstance(node, AnyWordSlot):
        # Any word at all -- which is every token the lexer can produce except
        # NUMBER, since a Defense name is made of words.
        return FAIL if token == NUMBER_TOKEN else EMPTY
    if isinstance(node, Opt):
        return derivative(node.part, token)
    if isinstance(node, Rep):
        # Having taken one, the rest are optional: d(P+) = d(P) . P*, and P*
        # is written Opt(Rep(P)) since there is no Star node.
        return _seq_of((derivative(node.part, token), Opt(node)))
    if isinstance(node, Seq):
        return _seq_derivative(node.parts, token)
    if isinstance(node, Alt):
        return _alt_of([derivative(o, token) for o in node.options])
    raise TypeError(node)


def _seq_derivative(parts: tuple[Node, ...], token: str) -> Node | _Empty | _Fail:
    if not parts:
        return FAIL
    head, *rest = parts
    branches: list[Node | _Empty | _Fail] = []
    d_head = derivative(head, token)
    if not isinstance(d_head, _Fail):
        branches.append(_seq_of((d_head, *rest)))
    if nullable(head):
        branches.append(_seq_derivative(tuple(rest), token))
    return _alt_of(branches)


def candidates(consumed_token_names: list[str]) -> set[str]:
    """The set of token names (or NUMBER_TOKEN, or ANY_WORD_TOKEN) valid
    immediately after having consumed `consumed_token_names`, per the grammar
    mirror."""
    node: Node | _Empty | _Fail = CALL
    for name in consumed_token_names:
        node = derivative(node, name)
        if isinstance(node, _Fail):
            return set()
    return first(node)


@dataclass(frozen=True)
class Suggestion:
    kind: str  # "keyword" | "number"
    label: str | None = None


def _find_caret_token_index(tokens: list, cursor: int) -> int:  # type: ignore[type-arg]
    for i, tok in enumerate(tokens):
        if cursor <= tok.stop + 1:
            return i
    return len(tokens) - 1


def suggest(text: str, cursor: int) -> list[Suggestion]:
    stream = CaseChangingCharStream(text, upper=True)
    lexer = CallsLexer(stream)
    token_stream = CommonTokenStream(lexer)
    token_stream.fill()
    all_tokens = token_stream.tokens

    caret_index = _find_caret_token_index(all_tokens, cursor)
    caret_token = all_tokens[caret_index]
    partial = text[caret_token.start : cursor].lower()
    partial_is_digits = partial.isdigit() or partial == ""

    consumed_names = [
        name
        for tok in all_tokens[:caret_index]
        if (name := CallsLexer.symbolicNames[tok.type]) is not None
    ]
    raw_candidates = candidates(consumed_names)

    suggestions: list[Suggestion] = []
    for name in raw_candidates:
        if name == NUMBER_TOKEN:
            if partial_is_digits:
                suggestions.append(Suggestion(kind="number"))
            continue
        # A Defense name is any word at all, so there is nothing here to
        # complete against -- the names the rulebook does list stand in.
        labels = DEFENSE_NAME_WORDS if name == ANY_WORD_TOKEN else [WORDS[name]]
        # Canonical words are capitalized; what the user has typed so far is
        # whatever they typed, lower-cased above -- so match case-insensitively
        # while still offering the canonical spelling.
        suggestions.extend(
            Suggestion(kind="keyword", label=label)
            for label in labels
            if label.lower().startswith(partial)
        )

    suggestions.sort(key=lambda s: (s.label is None, s.label or ""))
    return suggestions
